cpp/src/strings/convert/convert_urls.cu

```diff
@@ -185,137 +193,198 @@ constexpr bool is_hex_digit(char ch)
- * byte. For example, the sequence '%20' is converted into byte (0x20) which is a single space
- * character. Another example converts '%C3%A9' into 2 sequential bytes (0xc3 and 0xa9
- * respectively). Overall, 3 characters are converted into one byte whenever a '%' character
- * is encountered in the string.
+ * @param[in] in_chars Character buffer for the input string column.
+ * @param[in] in_offsets Offset value of each string associated with `in_chars`.
+ * @param[out] out_counts Number of characters in each decode URL.
+ * @param[in] num_strings Number of strings to count the number of decoded characters.
  */
-struct url_decode_char_replacer {
-  device_span<size_type const> const d_esc_positions{};
-  char const* const d_in_chars{};
-  char* const d_out_chars{};
-  int32_t const first_input_char_offset = 0;
-
-  /**
-   * @brief Copy an input character to the output, decoding escape sequences
-   *
-   * Each character position is examined against the list of known escape sequence positions.
-   * If the position is not within an escape sequence then the input character is copied to the
-   * output. If the position is the start of an escape sequence then the sequence is decoded to
-   * produce the character copied to the output. Any characters after the start of an escape
-   * sequence but still within the escape sequence are discarded.
-   *
-   * @param input_idx The input character position to process
-   */
-  __device__ void operator()(size_type input_idx) const
-  {
-    char ch = d_in_chars[input_idx];
-
-    // determine the number of escape sequences at or before this character position
-    size_type const* next_esc_pos_ptr =
-      thrust::upper_bound(thrust::seq, d_esc_positions.begin(), d_esc_positions.end(), input_idx);
-    size_type num_prev_esc = next_esc_pos_ptr - d_esc_positions.data();
-
-    // every escape that occurs before this one replaces 3 characters with 1
-    size_type output_idx = input_idx - (num_prev_esc * 2) - first_input_char_offset;
-    if (num_prev_esc > 0) {
-      size_type prev_esc_pos = *(next_esc_pos_ptr - 1);
-      // find the previous escape to see if this character is within the escape sequence
-      if (input_idx - prev_esc_pos < 3) {
-        if (input_idx == prev_esc_pos) {  // at a position that needs to be replaced
-          ch = (hex_char_to_byte(d_in_chars[input_idx + 1]) << 4) |
-               hex_char_to_byte(d_in_chars[input_idx + 2]);
-          // previous escape sequence is this position, so the original calculation over-adjusted
-          output_idx += 2;
-        } else {
-          // one of the escape hex digits that has no corresponding character in the output
-          return;
-        }
+template <int num_warps_per_threadblock, int char_block_size>
+__global__ void url_decode_char_counter(char const* const in_chars,
+                                        offset_type const* const in_offsets,
```

Comment:
Rather than passing the string children components here, you could pass a `column_device_view` of the input strings column and access each string using the `element<string_view>()` which handles all the offsets automatically.

Reply:
Addressed in https://github.com/rapidsai/cudf/pull/8622/commits/93497ab1aaf3688fdc2cb5fae1eb35de0636dcfd.